clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -3396,6 +3396,13 @@ LangAS CodeGenModule::GetGlobalVarAddressSpace(const VarDecl *D) {
     return AddrSpace;
   }
 
+  if (LangOpts.SYCLIsDevice) {
+    if (D && D->getType().isConstQualified())
+      return LangAS::opencl_constant;
+
+    return LangAS::opencl_global;
```

Comment:
We use `sycl_*` address spaces in all other places and with this change we mix two potentially different mappings in the single program.
I think we should first agree on using `opencl_*` address space map for SYCL device code and if so, replace all usages of `sycl_*` address spaces with `opencl_*` and remove `sycl_*` address spaces from the map.
@Naghasan, @keryell do you see any potential issues with using `opencl_*` address spaces?

Reply:
TBH, I never fully understood the rational behind the introduction of the `sycl_*` address spaces. I think it was made to allow this https://github.com/intel/llvm/commit/c27854310048c83a9f3edb60076e595f142f3f82#diff-54e8e44f96ab947511166d795bad8022R477 but I find this suspicious as it is easy to trick the compiler (see https://github.com/intel/llvm/issues/40).

Apart from the constant address space, I don't see any particular issue in moving from `sycl_*` asp to `opencl_*` asp.

The device compiler needs to convert the outlined code into an OpenCL kernel, so reusing `LangAS::opencl_*` seems a rational choice to me as it will play better with existing code in clang (especially that it will need to obey to the same set of rules). For the library, the exposure of the address space is pushed to the `multi_ptr` and `accessor` classes. You will probably have to do adjustments in the header files, but those classes are also there to handle the host/device boundary in a transparent way for the user.

That being said, there is the question of handling the (OpenCL) constant address space. As you can't make it overlap with the generic address space, this becomes tricky. But that's an exception that can be approached in different ways.

----

Note on the this patch itself

considering this code
```
static const int bar = 42;

void work(int*);

[...]
cgh.single_task([=]() {
  int foo = 21;
  work(&foo);
  work(&bar);
});
```
this patch will tranform it into something like this opencl C code:
```
constant const int bar = 42;

void work(generic int*);

__kernel void ker() {
  int foo = 21;
  work(&foo); // ok
  work(&bar); // error: constant and generic address space does not overlap
});
```

Reply:
>  // error: constant and generic address space does not overlap

Right, I noticed that as well. Please treat this patch as WIP for now.

Reply:
NOTE: even if we use `opencl_*` it doesn't mean we enforce OpenCL rules. A lot of OpenCL rules are enabled only for OpenCL language, which we can't use in for SYCL unfortunately.
We might need to go over this checks that enable them for SYCL too by adding additional case or removing language check at all.
+@AnastasiaStulova

Reply:
Btw,  I don't think it is correct to deduce **const** to **__constant**. Variables that are **const** qualified can be initialized at RT. But **__constant** required compile time constant initializers.

In a way it is a bit closer to **constexpr**.

> NOTE: even if we use opencl_* it doesn't mean we enforce OpenCL rules. A lot of OpenCL rules are enabled only for OpenCL language, which we can't use in for SYCL unfortunately.
> We might need to go over this checks that enable them for SYCL too by adding additional case or removing language check at all.
> +@AnastasiaStulova

I would prefer to remove OpenCL checks completely. They normally come from the old code anyway. I never understood why it was written like that. If we have OpenCL address spaces surely we are compiling OpenCL code (or may be SYCL device code now too!).

I was also thinking if it helps we could add some sort of common language mode (i.e. CLDevice) that we could use for common logic between OpenCL and SYCL. And then derive OpenCL and SYCL from CLDevice. Not sure how helpful it is though but it could be one options.